packages/client/src/service/skeleton.ts

```diff
@@ -1201,72 +1208,82 @@ export class Skeleton extends MetaDBManager {
 
       // Insert into chain
       let numBlocksInserted = 0
-      try {
-        numBlocksInserted = await this.chain.putBlocks([block], true)
-      } catch (e) {
-        this.config.logger.error(`fillCanonicalChain putBlock error=${(e as Error).message}`)
-        if (oldHead !== null && oldHead.header.number >= block.header.number) {
-          // Put original canonical head block back if reorg fails
-          // UPDATE
-          // not sure we can put oldHead because the oldHead chain might have been partially overwritten
-          // skipping for now, leaving code here for future cleanup/debugging
-          //
-          // await this.chain.putBlocks([oldHead], true)
-        }
-      }
-
-      // handle insertion failures
-      if (numBlocksInserted !== 1) {
-        this.config.logger.error(
-          `Failed to put block number=${number} fork=${block.common.hardfork()} hash=${short(
-            block.hash()
-          )} parentHash=${short(block.header.parentHash)}from skeleton chain to canonical`
-        )
-        // Lets log some parent by number and parent by hash, that may help to understand whats going on
-        let parent = null
+      let numBlocksSkipped = 0
+      // chain height has to be <= block number as we will skip putting this block as it might currently
+      // cause chain reset. This can happen if any other async process added a batch of blocks like
+      // execution's setHead. If that caused this chain to be not canonical anymore than the next
+      // putblocks should fail causing the fill to exit with skeleton stepback
+      if (this.chain.blocks.height <= block.header.number) {
```

Comment:
this is the only real change here, i.e. skip putting if chain has been advanced by some other putBlock process. But in the hive tests i ran locally, there would still be  a random race where this check would pass but still when actual writing the block, a concurrent putblock wins the race causing an issue 

(so we eventually needs to fix it since the only actual locking happens in the blockchain class)

Reply:
Maybe this is a silly question but do we need to check `block.header.number` and ensure its only one block ahead of `height`?

Reply:
if its not, the putBlock will fail anyway